File: codes/Garnetwzy/39.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ret;
        sort(candidates.begin(), candidates.end());
        vector<int> cur;
        dfs(ret, cur, 0, target, candidates, 0);
        return ret;
    }
    
    void dfs(vector<vector<int>>& ret, vector<int>& cur, int sum, int target, vector<int>& candidates, int index) {
        if(sum == target) {
            ret.push_back(cur);
            return;
        }
        if(sum > target)
            return;
        
        for(int i = index; i < candidates.size(); i++) {
            cur.push_back(candidates[i]);
            dfs(ret, cur, sum + candidates[i], target, candidates, i);
            cur.pop_back();
        }
    }
};

File: codes/HellDream/39.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> lists = new ArrayList<>();
        List<Integer> list = new ArrayList<>();
        combination(lists, list, candidates, target, 0, 0);
        return lists;
    }
    
    public void combination(List<List<Integer>> lists,List<Integer> list,int[] candidates, int target, int sum, int start){
        if(sum > target){
            return;
        }
        if(sum==target){
            lists.add(list);
            return;
        }
        for(int i=start;i<candidates.length;i++){
            if(sum+candidates[i]<=target){
                List<Integer> tmplist = new ArrayList<>(list);
                tmplist.add(candidates[i]);
                combination(lists, tmplist, candidates, target, sum+candidates[i],i);
            }
        }
    }
}